In [1]:
import pandas as pd
import os
from IPython.core.display import display
from datetime import datetime

print(os.listdir("."))

['quiz.ipynb', 'data.csv']


In [2]:
data = pd.read_csv('data.csv')

In [3]:
# Предобработка датасета

# Дата выхода на экраны
data['release_datetime'] = data['release_date'].apply(lambda v: datetime.strptime(v, '%m/%d/%Y'))

# Месяц выхода на экраны
data['release_month'] = data['release_datetime'].apply(lambda d: d.month)

def split(source):
    return [e.strip() for e in source.split(sep='|')]

# Список актёров
data['actors'] = data['cast'].apply(split)

# Cписок режисёров
data['directors'] = data['director'].apply(split)

# Список жанров
data['genres_list'] = data['genres'].apply(split)

# Список компаний
data['production_companies_list'] = data['production_companies'].apply(split)

# Прибыль
data['gross_profit'] = data['revenue'] - data['budget']

# Прибыль на каждого режисёра
data['gross_profit_by_director'] = data['gross_profit'] / data['directors'].apply(len)

# Прибыль на каждого режисёра
data['gross_profit_by_actor'] = data['gross_profit'] / data['actors'].apply(len)

# Прибыль на компанию
data['gross_profit_by_production_company'] = data['gross_profit'] / data['production_companies_list'].apply(len)

display(data)

imdb_id  popularity     budget     revenue  \
0     tt0369610   32.985763  150000000  1513528810   
1     tt1392190   28.419936  150000000   378436354   
2     tt2908446   13.112507  110000000   295238201   
3     tt2488496   11.173104  200000000  2068178225   
4     tt2820852    9.335014  190000000  1506249360   
...         ...         ...        ...         ...   
1885  tt0120903    0.182246   75000000   157299717   
1886  tt0192255    0.123920   22000000    13555988   
1887  tt0131704    0.104908   76000000    35134820   
1888  tt0162983    0.031018   40000000    36037909   
1889  tt0163676    0.015997   15000000     5217498   

                            original_title  \
0                           Jurassic World   
1                       Mad Max: Fury Road   
2                                Insurgent   
3             Star Wars: The Force Awakens   
4                                Furious 7   
...                                    ...   
1885                                 X-Men   
1886                    The Little Vampire   
1887  The Adventures of Rocky & Bullwinkle   
1888                            Hanging Up   
1889                          The In Crowd   

                                                   cast          director  \
0     Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   Colin Trevorrow   
1     Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...     George Miller   
2     Shailene Woodley|Theo James|Kate Winslet|Ansel...  Robert Schwentke   
3     Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...       J.J. Abrams   
4     Vin Diesel|Paul Walker|Jason Statham|Michelle ...         James Wan   
...                                                 ...               ...   
1885  Patrick Stewart|Hugh Jackman|Ian McKellen|Hall...      Bryan Singer   
1886  Richard E. Grant|Jonathan Lipnicki|Jim Carter|...          Uli Edel   
1887  Rene Russo|Jason Alexander|Piper Perabo|Randy ...       Des McAnuff   
1888  Meg Ryan|Diane Keaton|Lisa Kudrow|Walter Matth...      Diane Keaton   
1889  Susan Ward|Lori Heuring|Matthew Settle|Nathan ...      Mary Lambert   

                                                tagline  \
0                                     The park is open.   
1                                    What a Lovely Day.   
2                            One Choice Can Destroy You   
3                         Every generation has a story.   
4                                   Vengeance Hits Home   
...                                                 ...   
1885                                   Evolution Begins   
1886  They're not just best friends, they're blood b...   
1887            This summer it's not the same old bull.   
1888                   Every family has a few hang-ups.   
1889                       What would you do to get in?   

                                               overview  runtime  ...  \
0     Twenty-two years after the events of Jurassic ...      124  ...   
1     An apocalyptic story set in the furthest reach...      120  ...   
2     Beatrice Prior must confront her inner demons ...      119  ...   
3     Thirty years after defeating the Galactic Empi...      136  ...   
4     Deckard Shaw seeks revenge against Dominic Tor...      137  ...   
...                                                 ...      ...  ...   
1885  Two mutants, Rogue and Wolverine, come to a pr...      104  ...   
1886  Based on the popular books, the story tells of...       95  ...   
1887  Rocky and Bullwinkle have been living off the ...       88  ...   
1888  A trio of sisters bond over their ambivalence ...       94  ...   
1889  A mentally disturbed young woman takes a job a...      105  ...   

     release_datetime release_month  \
0          2015-06-09             6   
1          2015-05-13             5   
2          2015-03-18             3   
3          2015-12-15            12   
4          2015-04-01             4   
...               ...           ...   
1885       2000-07-13 

In [4]:
answer_ls = [] # создадим список с ответами. сюда будем добавлять ответы по мере прохождения теста

# 1. У какого фильма из списка самый большой бюджет?
Варианты ответов:
1. The Dark Knight Rises (tt1345836)
2. Spider-Man 3 (tt0413300)
3. Avengers: Age of Ultron (tt2395427)
4. The Warrior's Way	(tt1032751)
5. Pirates of the Caribbean: On Stranger Tides (tt1298650)

In [5]:
display(data.sort_values(by=['budget'], ascending=False).head(1))

imdb_id  popularity     budget   revenue     original_title  \
491  tt1032751     0.25054  425000000  11087569  The Warrior's Way   

                                                  cast    director  \
491  Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...  Sngmoo Lee   

                     tagline  \
491  Assassin. Hero. Legend.   

                                              overview  runtime  ...  \
491  An Asian assassin (Dong-gun Jang) is forced to...      100  ...   

    release_datetime release_month  \
491       2010-12-02            12   

                                                actors     directors  \
491  [Kate Bosworth, Jang Dong-gun, Geoffrey Rush, ...  [Sngmoo Lee]   

                                         genres_list  \
491  [Adventure, Fantasy, Action, Western, Thriller]   

      production_companies_list gross_profit  gross_profit_by_director  \
491  [Boram Entertainment Inc.]   -413912431              -413912431.0   

    gross_profit_by_actor gross_profit_by_production_company  
491           -82782486.2                       -413912431.0  

[1 rows x 26 columns]

In [6]:
answer_ls.append(4)

# 2. Какой из фильмов самый длительный (в минутах)?
1. The Lord of the Rings: The Return of the King	(tt0167260)
2. Gods and Generals	(tt0279111)
3. King Kong	(tt0360717)
4. Pearl Harbor	(tt0213149)
5. Alexander	(tt0346491)

In [7]:
display(data.sort_values(by=['runtime'], ascending=False).head(1))

imdb_id  popularity    budget   revenue     original_title  \
1158  tt0279111    0.469518  56000000  12923936  Gods and Generals   

                                                   cast           director  \
1158  Stephen Lang|Jeff Daniels|Robert Duvall|Kevin ...  Ronald F. Maxwell   

                                  tagline  \
1158  The nations heart was touched by...   

                                               overview  runtime  ...  \
1158  The film centers mostly around the personal an...      214  ...   

     release_datetime release_month  \
1158       2003-02-21             2   

                                                 actors            directors  \
1158  [Stephen Lang, Jeff Daniels, Robert Duvall, Ke...  [Ronald F. Maxwell]   

                genres_list              production_companies_list  \
1158  [Drama, History, War]  [Turner Pictures, Antietam Filmworks]   

     gross_profit  gross_profit_by_director gross_profit_by_actor  \
1158    -43076064               -43076064.0            -8615212.8   

     gross_profit_by_production_company  
1158                        -21538032.0  

[1 rows x 26 columns]

In [8]:
answer_ls.append(2)

# 3. Какой из фильмов самый короткий (в минутах)?
Варианты ответов:

1. Home on the Range	tt0299172
2. The Jungle Book 2	tt0283426
3. Winnie the Pooh	tt1449283
4. Corpse Bride	tt0121164
5. Hoodwinked!	tt0443536

In [9]:
display(data.sort_values(by=['runtime'], ascending=True).head(1))

imdb_id  popularity    budget   revenue   original_title  \
769  tt1449283    1.425344  30000000  14460000  Winnie the Pooh   

                                                  cast  \
769  Jim Cummings|Travis Oates|Jim Cummings|Bud Luc...   

                      director   tagline  \
769  Stephen Anderson|Don Hall  Oh Pooh.   

                                              overview  runtime  ...  \
769  During an ordinary day in Hundred Acre Wood, W...       63  ...   

    release_datetime release_month  \
769       2011-04-13             4   

                                                actors  \
769  [Jim Cummings, Travis Oates, Jim Cummings, Bud...   

                        directors          genres_list  \
769  [Stephen Anderson, Don Hall]  [Animation, Family]   

                             production_companies_list gross_profit  \
769  [Walt Disney Pictures, Walt Disney Animation S...    -15540000   

     gross_profit_by_director gross_profit_by_actor  \
769                -7770000.0            -3108000.0   

    gross_profit_by_production_company  
769                         -7770000.0  

[1 rows x 26 columns]

In [10]:
answer_ls.append(3)

# 4. Какое число ближе к средней длительности фильма в датасете?
Варианты ответов:
1. 115
2. 110
3. 105
4. 120
5. 100

In [11]:
print(round(data['runtime'].mean()))

110


In [12]:
answer_ls.append(2)

# 5. Какое число ближе к медианной длительности фильма в датасете?
Варианты ответов:
1. 106
2. 112
3. 101
4. 120
5. 115


In [13]:
print(round(data['runtime'].median()))

106


In [14]:
answer_ls.append(1)

# 6. Какой самый прибыльный фильм?
> Внимание! Здесь и далее под «прибылью» или «убытками» понимается разность между сборами и бюджетом фильма.
Варианты ответов:
1. The Avengers	tt0848228
2. Minions	tt2293640
3. Star Wars: The Force Awakens	tt2488496
4. Furious 7	tt2820852
5. Avatar	tt0499549

In [15]:
display(data.sort_values(by=['gross_profit'], ascending=False).head(1))

imdb_id  popularity     budget     revenue original_title  \
239  tt0499549    9.432768  237000000  2781505847         Avatar   

                                                  cast       director  \
239  Sam Worthington|Zoe Saldana|Sigourney Weaver|S...  James Cameron   

                         tagline  \
239  Enter the World of Pandora.   

                                              overview  runtime  ...  \
239  In the 22nd century, a paraplegic Marine is di...      162  ...   

    release_datetime release_month  \
239       2009-12-10            12   

                                                actors        directors  \
239  [Sam Worthington, Zoe Saldana, Sigourney Weave...  [James Cameron]   

                                       genres_list  \
239  [Action, Adventure, Fantasy, Science Fiction]   

                             production_companies_list gross_profit  \
239  [Ingenious Film Partners, Twentieth Century Fo...   2544505847   

     gross_profit_by_director gross_profit_by_actor  \
239              2.544506e+09           508901169.4   

    gross_profit_by_production_company  
239                       6.361265e+08  

[1 rows x 26 columns]

In [16]:
answer_ls.append(5)

# 7. Какой фильм самый убыточный?
Варианты ответов:
1. Supernova tt0134983
2. The Warrior's Way tt1032751
3. Flushed Away	tt0424095
4. The Adventures of Pluto Nash	tt0180052
5. The Lone Ranger	tt1210819

In [17]:
display(data.sort_values(by=['gross_profit'], ascending=True).head(1))

imdb_id  popularity     budget   revenue     original_title  \
491  tt1032751     0.25054  425000000  11087569  The Warrior's Way   

                                                  cast    director  \
491  Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...  Sngmoo Lee   

                     tagline  \
491  Assassin. Hero. Legend.   

                                              overview  runtime  ...  \
491  An Asian assassin (Dong-gun Jang) is forced to...      100  ...   

    release_datetime release_month  \
491       2010-12-02            12   

                                                actors     directors  \
491  [Kate Bosworth, Jang Dong-gun, Geoffrey Rush, ...  [Sngmoo Lee]   

                                         genres_list  \
491  [Adventure, Fantasy, Action, Western, Thriller]   

      production_companies_list gross_profit  gross_profit_by_director  \
491  [Boram Entertainment Inc.]   -413912431              -413912431.0   

    gross_profit_by_actor gross_profit_by_production_company  
491           -82782486.2                       -413912431.0  

[1 rows x 26 columns]

In [18]:
answer_ls.append(5)

# 8. У скольких фильмов из датасета объем сборов оказался выше бюджета?
Варианты ответов:
1. 1478
2. 1520
3. 1241
4. 1135
5. 1398


In [19]:
print(len(data[(data['gross_profit'] > 0)]))

1478


In [20]:
answer_ls.append(1)

# 9. Какой фильм оказался самым кассовым в 2008 году?
Варианты ответов:
1. Madagascar: Escape 2 Africa	tt0479952
2. Iron Man	tt0371746
3. Kung Fu Panda	tt0441773
4. The Dark Knight	tt0468569
5. Mamma Mia!	tt0795421

In [21]:
display(data[(data['release_year'] == 2008)].sort_values(by=['revenue'], ascending=False).head(1))

imdb_id  popularity     budget     revenue   original_title  \
600  tt0468569    8.466668  185000000  1001921825  The Dark Knight   

                                                  cast           director  \
600  Christian Bale|Michael Caine|Heath Ledger|Aaro...  Christopher Nolan   

             tagline                                           overview  \
600  Why So Serious?  Batman raises the stakes in his war on crime. ...   

     runtime  ... release_datetime release_month  \
600      152  ...       2008-07-16             7   

                                                actors            directors  \
600  [Christian Bale, Michael Caine, Heath Ledger, ...  [Christopher Nolan]   

                          genres_list  \
600  [Drama, Action, Crime, Thriller]   

                             production_companies_list gross_profit  \
600  [DC Comics, Legendary Pictures, Warner Bros., ...    816921825   

     gross_profit_by_director gross_profit_by_actor  \
600               816921825.0           163384365.0   

    gross_profit_by_production_company  
600                       2.042305e+08  

[1 rows x 26 columns]

In [22]:
answer_ls.append(4)

# 10. Самый убыточный фильм за период с 2012 по 2014 гг. (включительно)?
Варианты ответов:
1. Winter's Tale	tt1837709
2. Stolen	tt1656186
3. Broken City	tt1235522
4. Upside Down	tt1374992
5. The Lone Ranger	tt1210819


In [23]:
display(data[(2012 <= data['release_year']) & (data['release_year'] <= 2014)].sort_values(by=['gross_profit'], ascending=True).head(1))

imdb_id  popularity     budget   revenue   original_title  \
1246  tt1210819     1.21451  255000000  89289910  The Lone Ranger   

                                                   cast        director  \
1246  Johnny Depp|Armie Hammer|William Fichtner|Hele...  Gore Verbinski   

                      tagline  \
1246  Never Take Off the Mask   

                                               overview  runtime  ...  \
1246  The Texas Rangers chase down a gang of outlaws...      149  ...   

     release_datetime release_month  \
1246       2013-07-03             7   

                                                 actors         directors  \
1246  [Johnny Depp, Armie Hammer, William Fichtner, ...  [Gore Verbinski]   

                       genres_list  \
1246  [Action, Adventure, Western]   

                              production_companies_list gross_profit  \
1246  [Walt Disney Pictures, Jerry Bruckheimer Films...   -165710090   

      gross_profit_by_director gross_profit_by_actor  \
1246              -165710090.0           -33142018.0   

     gross_profit_by_production_company  
1246                        -33142018.0  

[1 rows x 26 columns]

In [24]:
answer_ls.append(1)

# 11. Какого жанра фильмов больше всего?
Варианты ответов:
1. Action
2. Adventure
3. Drama
4. Comedy
5. Thriller

In [25]:
display(data['genres_list'].explode().value_counts().head(1))

Drama    782
Name: genres_list, dtype: int64

In [26]:
answer_ls.append(1)

# 12. Какого жанра среди прибыльных фильмов больше всего?
Варианты ответов:
1. Drama
2. Comedy
3. Action
4. Thriller
5. Adventure

In [27]:
display(data[(data['gross_profit'] > 0)]['genres_list'].explode().value_counts().head(1))

Drama    560
Name: genres_list, dtype: int64

In [28]:
answer_ls.append(1)

# 13. Кто из режиссеров снял больше всего фильмов?
Варианты ответов:
1. Steven Spielberg
2. Ridley Scott 
3. Steven Soderbergh
4. Christopher Nolan
5. Clint Eastwood

In [29]:
display(data['directors'].explode().value_counts().head(1))

Steven Soderbergh    13
Name: directors, dtype: int64

In [30]:
answer_ls.append(4)

# 14. Кто из режиссеров снял больше всего прибыльных фильмов?
Варианты ответов:
1. Steven Soderbergh
2. Ridley Scott
3. Steven Spielberg
4. Christopher Nolan
5. Clint Eastwood

In [31]:
only_profitable_movies = data[(data['gross_profit'] > 0)]
display(only_profitable_movies['directors'].explode().value_counts().head(1))

Ridley Scott    12
Name: directors, dtype: int64

In [32]:
answer_ls.append(2)

# 15. Кто из режиссеров принес больше всего прибыли?
Варианты ответов:
1. Steven Spielberg
2. Christopher Nolan
3. David Yates
4. James Cameron
5. Peter Jackson


In [33]:
display(data.explode('directors').groupby('directors').sum().sort_values(by=['gross_profit_by_director'], ascending=False).head(1))

popularity      budget     revenue  runtime  vote_count  \
directors                                                                
Peter Jackson   45.679675  1288000000  6490593685     1355       31197   

               vote_average  release_year  release_month  gross_profit  \
directors                                                                
Peter Jackson          58.1         16059             95    5202593685   

               gross_profit_by_director  gross_profit_by_actor  \
directors                                                        
Peter Jackson              5.202594e+09           1.040519e+09   

               gross_profit_by_production_company  
directors                                          
Peter Jackson                        1.813866e+09

In [34]:
answer_ls.append(5)

# 16. Какой актер принес больше всего прибыли?
Варианты ответов:
1. Emma Watson
2. Johnny Depp
3. Michelle Rodriguez
4. Orlando Bloom
5. Rupert Grint

In [35]:
display(data.explode('actors').groupby('actors').sum().sort_values(by=['gross_profit_by_actor'], ascending=False).head(1))

popularity      budget     revenue  runtime  vote_count  \
actors                                                                 
Emma Watson   52.428017  1493000000  8159245597     1609       32702   

             vote_average  release_year  release_month  gross_profit  \
actors                                                                 
Emma Watson          82.8         24096             96    6666245597   

             gross_profit_by_director  gross_profit_by_actor  \
actors                                                         
Emma Watson              6.670807e+09           1.333249e+09   

             gross_profit_by_production_company  
actors                                           
Emma Watson                        2.246273e+09

In [36]:
answer_ls.append(1)

# 17. Какой актер принес меньше всего прибыли в 2012 году?
Варианты ответов:
1. Nicolas Cage
2. Danny Huston
3. Kirsten Dunst
4. Jim Sturgess
5. Sami Gayle

In [37]:
movies_2012 = data[(data['release_year'] == 2012)]
display(movies_2012.explode('actors').groupby('actors').sum().sort_values(by=['gross_profit_by_actor'], ascending=True).head(1))

popularity    budget   revenue  runtime  vote_count  \
actors                                                               
Kirsten Dunst     2.24286  85000000  16890793      241         900   

               vote_average  release_year  release_month  gross_profit  \
actors                                                                   
Kirsten Dunst          11.7          4024             14     -68109207   

               gross_profit_by_director  gross_profit_by_actor  \
actors                                                           
Kirsten Dunst               -68109207.0            -13621841.4   

               gross_profit_by_production_company  
actors                                             
Kirsten Dunst                         -13621841.4

In [38]:
answer_ls.append(3)

# 18. Какой актер снялся в большем количестве высокобюджетных фильмов? Примечание: в фильмах, где бюджет выше среднего по данной выборке.
Варианты ответов:
1. Tom Cruise
2. Mark Wahlberg 
3. Matt Damon
4. Angelina Jolie
5. Adam Sandler

In [39]:
mean_movie_budget = data['budget'].mean()

high_budget_movies = data[(data['budget'] > mean_movie_budget)]
display(data[(data['budget'] > mean_movie_budget)].explode('actors')['actors'].value_counts().head(1))

Matt Damon    18
Name: actors, dtype: int64

In [40]:
answer_ls.append(3)

# 19. В фильмах какого жанра больше всего снимался Nicolas Cage?
Варианты ответа:
1. Drama
2. Action
3. Thriller
4. Adventure
5. Crime

In [41]:
cage_movies = data[data['actors'].apply(lambda actors: 'Nicolas Cage' in actors)]

display(cage_movies.explode('genres_list')['genres_list'].value_counts().head(1))

Action    17
Name: genres_list, dtype: int64

In [42]:
answer_ls.append(2)

# 20. Какая студия сняла больше всего фильмов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Paramount Pictures
3. Columbia Pictures
4. Warner Bros
5. Twentieth Century Fox Film Corporation

In [43]:
display(data.explode('production_companies_list')['production_companies_list'].value_counts().head(1))

Universal Pictures    173
Name: production_companies_list, dtype: int64

In [44]:
answer_ls.append(1)

# 21. Какая студия сняла больше всего фильмов в 2015 году?
Варианты ответа:
1. Universal Pictures
2. Paramount Pictures
3. Columbia Pictures
4. Warner Bros
5. Twentieth Century Fox Film Corporation

In [45]:
movies_2015 = data[(data['release_year'] == 2015)]

display(movies_2015.explode('production_companies_list')['production_companies_list'].value_counts().head(1))

Warner Bros.    12
Name: production_companies_list, dtype: int64

In [46]:
answer_ls.append(4)

# 22. Какая студия заработала больше всего денег в жанре комедий за все время?
Варианты ответа:
1. Warner Bros
2. Universal Pictures (Universal)
3. Columbia Pictures
4. Paramount Pictures
5. Walt Disney

In [47]:
comedy_movies = data[data['genres_list'].apply(lambda genres: 'Comedy' in genres)]

display(comedy_movies.explode('production_companies_list')[['production_companies_list', 'gross_profit_by_production_company']].groupby('production_companies_list').sum().sort_values(by=['gross_profit_by_production_company'], ascending=False).head(1))

gross_profit_by_production_company
production_companies_list                                    
Walt Disney Pictures                             3.660158e+09

In [48]:
display(comedy_movies.explode('production_companies_list')[['production_companies_list', 'gross_profit']].groupby('production_companies_list').sum().sort_values(by=['gross_profit'], ascending=False).head(1))

gross_profit
production_companies_list              
Universal Pictures           8961545581

In [49]:
answer_ls.append(2)

# 23. Какая студия заработала больше всего денег в 2012 году?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Columbia Pictures
4. Paramount Pictures
5. Lucasfilm

In [50]:
display(movies_2012.explode('production_companies_list')[['production_companies_list', 'gross_profit']].groupby('production_companies_list').sum().sort_values(by=['gross_profit'], ascending=False).head(1))

gross_profit
production_companies_list              
Columbia Pictures            2501406608

In [51]:
answer_ls.append(3)

# 24. Самый убыточный фильм от Paramount Pictures?
Варианты ответа:

1. K-19: The Widowmaker tt0267626
2. Next tt0435705
3. Twisted tt0315297
4. The Love Guru tt0811138
5. The Fighter tt0964517

In [52]:
paramount_movies = data[data['production_companies_list'].apply(lambda companies: 'Paramount Pictures' in companies)]
display(paramount_movies[['original_title', 'gross_profit']].sort_values(by=['gross_profit'], ascending=True).head(1))

original_title  gross_profit
926  K-19: The Widowmaker     -64831034

In [53]:
answer_ls.append(1)

# 25. Какой самый прибыльный год (в какой год студии заработали больше всего)?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2002
5. 2015

In [54]:
display(data[['release_year', 'gross_profit']].groupby(['release_year']).sum().sort_values(by=['gross_profit'], ascending=False).head(1))

gross_profit
release_year              
2015           18668572378

In [55]:
answer_ls.append(5)

# 26. Какой самый прибыльный год для студии Warner Bros?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2010
5. 2015

In [56]:
warner_movies = data[data['production_companies'].apply(lambda companies: 'Warner Bros' in companies)]
display(warner_movies[['release_year', 'gross_profit']].groupby(['release_year']).sum().sort_values(by=['gross_profit'], ascending=False).head(1))

gross_profit
release_year              
2014            2295464519

In [57]:
answer_ls.append(1)

# 27. В каком месяце за все годы суммарно вышло больше всего фильмов?
Варианты ответа:
1. Январь
2. Июнь
3. Декабрь
4. Сентябрь
5. Май

In [58]:
display(data['release_month'].value_counts().head(1))

9    227
Name: release_month, dtype: int64

In [59]:
answer_ls.append(4)

# 28. Сколько суммарно вышло фильмов летом (за июнь, июль, август)?
Варианты ответа:
1. 345
2. 450
3. 478
4. 523
5. 381

In [60]:
len(data[(data['release_month'].isin([6, 7, 8]))])

450

In [61]:
answer_ls.append(2)

# 29. Какой режиссер выпускает (суммарно по годам) больше всего фильмов зимой?
Варианты ответов:
1. Steven Soderbergh
2. Christopher Nolan
3. Clint Eastwood
4. Ridley Scott
5. Peter Jackson

In [62]:
display(data[(data['release_month'].isin([12,1,2]))][['directors', 'release_month']].explode('directors')['directors'].value_counts().head(1))

Peter Jackson    7
Name: directors, dtype: int64

In [63]:
answer_ls.append(5)

# 30. Какой месяц чаще всего самый прибыльный в году?
> Почему нужно считать по годам, а не просто взять самый прибыльный месяц за весь период? Киноиндустрия активно развивается, прибыль компаний растет с каждым годом (например, в 2014 г. индустрия заработала в 2 раза больше, чем в 2004). Поэтому прибыльность более поздних месяцев будет значительно перевешивать предыдущие, что может привести к искажению выводов.
Варианты ответа:
1. Январь
2. Июнь
3. Декабрь
4. Сентябрь
5. Май

In [64]:
profit_by_month = data[['release_year', 'release_month', 'gross_profit']].groupby(['release_year', 'release_month']).sum()

profit_by_year = data[['release_year', 'gross_profit']].groupby(['release_year']).sum()

profit_by_year_month = profit_by_month.merge(profit_by_year, on=['release_year'], right_index=True, suffixes=('_by_month', '_by_year'))
profit_by_year_month['relative_month_profit'] = profit_by_year_month['gross_profit_by_month'] / profit_by_year_month['gross_profit_by_year']

relative_profit_by_month = profit_by_year_month.reset_index()[['release_month', 'relative_month_profit']].groupby(['release_month']).sum()

display(relative_profit_by_month.sort_values(by=['relative_month_profit'], ascending=False).head(1))

relative_month_profit
release_month                       
12                          2.356393

In [65]:
answer_ls.append(3)

# 31. Названия фильмов какой студии в среднем самые длинные по количеству символов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. Four By Two Productions

In [66]:
data['original_title_lenght'] = data['original_title'].apply(len)

movie_by_company = data[['original_title_lenght', 'production_companies_list']].explode('production_companies_list')

display(movie_by_company.groupby(['production_companies_list']).mean().sort_values(by=['original_title_lenght'], ascending=False).head(1))

original_title_lenght
production_companies_list                       
Four By Two Productions                     83.0

In [67]:
answer_ls.append(5)

# 32. Названия фильмов какой студии в среднем самые длинные по количеству слов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. Four By Two Productions

In [68]:
data['original_title_word_count'] = data['original_title'].apply(lambda t: len(t.split()))

movie_by_company = data[['original_title_word_count', 'production_companies_list']].explode('production_companies_list')

display(movie_by_company.groupby(['production_companies_list']).mean().sort_values(by=['original_title_word_count'], ascending=False).head(1))

original_title_word_count
production_companies_list                           
Four By Two Productions                         12.0

In [69]:
answer_ls.append(5)

# 33. Сколько разных слов используется в названиях фильмов (без учета регистра)?
Варианты ответа:
1. 6540
2. 1002
3. 2461
4. 28304
5. 3432

In [70]:
len(data['original_title'].apply(lambda t: [w.upper() for w in t.split()]).explode().unique())

2461

In [71]:
answer_ls.append(3)

# 34. Какие фильмы входят в 1 % лучших по рейтингу?
Варианты ответа:
1. Inside Out, Gone Girl, 12 Years a Slave
2. BloodRayne, The Adventures of Rocky & Bullwinkle
3. The Lord of the Rings: The Return of the King
4. 300, Lucky Number Slevin

In [72]:
display(data[['original_title', 'vote_average']][data['vote_average'] > data.quantile(0.99, numeric_only=True)['vote_average']])

original_title  vote_average
9                                        Inside Out           8.0
34                                             Room           8.0
118                                    Interstellar           8.0
119                         Guardians of the Galaxy           7.9
125                              The Imitation Game           8.0
128                                       Gone Girl           7.9
138                        The Grand Budapest Hotel           7.9
370                                       Inception           7.9
600                                 The Dark Knight           8.1
873                                     The Pianist           7.9
1082  The Lord of the Rings: The Return of the King           7.9
1184                        The Wolf of Wall Street           7.9
1192                               12 Years a Slave           7.9
1801                                        Memento           7.9

In [73]:
answer_ls.append(1)

# 35. Какие актеры чаще всего снимаются в одном фильме вместе?
Варианты ответа:
1. Johnny Depp & Helena Bonham Carter
2. Hugh Jackman & Ian McKellen
3. Vin Diesel & Paul Walker
4. Adam Sandler & Kevin James
5. Daniel Radcliffe & Rupert Grint

In [74]:
actor_by_movie = data[['actors', 'original_title']].explode('actors')

actor_pairs = actor_by_movie.merge(
    actor_by_movie, on=['original_title'], suffixes=('_left', '_right')
)[['actors_left', 'actors_right']].query(
    'actors_left != actors_right'
)

def build_a_pair(l):
    l.sort()
    return ' & '.join(l)

actor_pairs['pair'] = actor_pairs[['actors_left', 'actors_right']].values.tolist()
display(actor_pairs['pair'].apply(build_a_pair).value_counts().head(5))

Daniel Radcliffe & Rupert Grint       16
Emma Watson & Rupert Grint            16
Daniel Radcliffe & Emma Watson        16
Ben Stiller & Owen Wilson             12
Helena Bonham Carter & Johnny Depp    12
Name: pair, dtype: int64

In [75]:
answer_ls.append(5)

# 36. У какого из режиссеров самый высокий процент фильмов со сборами выше бюджета?
Варианты ответа:
1. Quentin Tarantino
2. Steven Soderbergh
3. Robert Rodriguez
4. Christopher Nolan
5. Clint Eastwood

In [76]:
gross_profit_by_director = data[['directors', 'gross_profit']].explode('directors')
gross_profit_by_director = gross_profit_by_director[
    (
        gross_profit_by_director['directors'].isin(
            [
                'Quentin Tarantino',
                'Steven Soderbergh',
                'Robert Rodriguez',
                'Christopher Nolan',
                'Clint Eastwood',
            ]
        )
    )
]


gross_profit_by_director['profitable'] = gross_profit_by_director['gross_profit'].apply(lambda p: p > 0)

successes = gross_profit_by_director[['directors', 'profitable']].groupby(
    ['directors']
).agg(
    ['sum', 'count']
)['profitable']

successes['percent'] = successes['sum'] / successes['count']
display(successes.sort_values(by=['percent'], ascending=False).head(2))

sum  count   percent
directors                              
Christopher Nolan  8.0      8  1.000000
Quentin Tarantino  6.0      7  0.857143

In [77]:
answer_ls.append(4)

# Submission

In [78]:
len(answer_ls)

36

In [79]:
pd.DataFrame({'Answer':answer_ls}, index=range(1,len(answer_ls)+1))

Answer
1        4
2        2
3        3
4        2
5        1
6        5
7        5
8        1
9        4
10       1
11       1
12       1
13       4
14       2
15       5
16       1
17       3
18       3
19       2
20       1
21       4
22       2
23       3
24       1
25       5
26       1
27       4
28       2
29       5
30       3
31       5
32       5
33       3
34       1
35       5
36       4